src/util/cluster.js

```diff
@@ -1,122 +1,136 @@
+ * true up to the specified maximum number of respawns, after which it will return false.
+ *
+ * @param {number} maxStuckWorkerRespawns the maximum number of times a stuck
+ * worker can be respawned
+ */
+function stuckWorkerRespawnFunc(maxStuckWorkerRespawns) {
+  let stuckWorkerRespawns = 0; // Counter for stuck worker respawns
+  return () => {
+    if (stuckWorkerRespawns < maxStuckWorkerRespawns) {
+      stuckWorkerRespawns += 1;
+      return true; // Indicate that the worker should be respawned
+    }
+    return false; // Do not respawn the worker if the limit is reached
+  };
 }
 
-function start(port, app, metricsApp) {
+/**
+ * Start the cluster manager.
+ *
+ * The following environment variables can be set to configure the cluster manager:
+ * - `NUM_PROCS`: Number of worker processes to spawn (default: 1)
+ * - `CLUSTER_MANAGER_SHUTDOWN_TIMEOUT`: Maximum time in ms to wait for graceful shutdown (default: 30000)
+ * - `CLUSTER_MANAGER_MAX_KILLED_WORKER_RESTARTS`: Maximum number of times a worker that dies unexpectedly will be restarted (default: 3)
+ * - `CLUSTER_MANAGER_MAX_STUCK_WORKER_RESPAWNS`: Maximum number of times stuck workers will be respawned (default: 3)
+ * - `STATS_CLIENT`: If set to 'prometheus', a metrics app will be started on a different port METRICS_PORT (default: 9091).
+ *
+ * @param {*} port the port to start the app on
+ * @param {*} app the app instance
+ * @param {*} metricsApp the app instance for metrics, it will be started on a different port (METRICS_PORT)
+ */
+async function start(port, app, metricsApp) {
   // Enable diagnostic reporting on SIGUSR2 signal for debugging stuck processes
   process.report.reportOnSignal = true;
 
-  if (cluster.isMaster) {
-    logger.info(`Master (pid: ${process.pid}) has started`);
-
-    // HTTP server for exposing metrics
-    if (process.env.STATS_CLIENT === 'prometheus') {
-      const metricsServer = metricsApp.listen(metricsPort);
-
-      gracefulShutdown(metricsServer, {
-        signals: 'SIGINT SIGTERM SIGSEGV',
-        timeout: 30000, // timeout: 30 secs
-        forceExit: false, // triggers process.exit() at the end of shutdown process
-      });
-    }
-
-    // Fork workers.
-    for (let i = 0; i < numWorkers; i += 1) {
-      cluster.fork({
-        WORKER_ID: `worker-${i + 1}`,
-      });
-    }
-
-    cluster.on('online', (worker) => {
-      logger.info(`Worker (pid: ${worker.process.pid}) is online`);
-      // To provide caching at pod-level
-    });
-
-    let isShuttingDown = false;
-    cluster.on('exit', async (worker) => {
-      if (!isShuttingDown) {
-        logger.error(`Worker (pid: ${worker.process.pid}) died`);
-        logger.error(`Killing other workers to avoid any side effects of the dead worker`);
-        logProcessInfo();
-        isShuttingDown = true;
-        await shutdownWorkers();
+  const numWorkers = parseInt(process.env.NUM_PROCS || '1', 10);
```

Comment:
https://github.com/rudderlabs/rudder-transformer/blob/develop/.github/workflows/verify-server-start.yml Currently this tests server is stating and responding, can we add few steps to test it is working correct when NUM_PROCS=4 or greater than 1.